Implement Send + Sync for VarLenUnicode, VarLenAscii, VarLenArray

I am implementing a utility to write data to an HDF5 file in a thread, which works fine for some types but not others. For example, any type that includes `VarLenUnicode`, `VarLenAscii` or `VarLenArray` cannot be sent between threads.

Here is a minimal example:
```rust
use hdf5::types::VarLenUnicode;

#[derive(hdf5::H5Type)]
#[repr(C)]
struct CustomType {
    text: VarLenUnicode,
}

fn write_in_thread() {
    let file = hdf5::File::create("/tmp/dset.h5").unwrap();

    let data = vec![
        CustomType {
            text: VarLenUnicode::from_str("foo").unwrap(),
        },
        CustomType {
            text: VarLenUnicode::from_str("bar").unwrap(),
        },
    ];

    std::thread::spawn(move || {
        file.group("/")
            .unwrap()
            .new_dataset_builder()
            .with_data(&data)
            .create("dataset")
            .unwrap();
    });
}
```
Output (truncated):
```
error[E0277]: `*mut u8` cannot be sent between threads safely
   --> src\io\src\writer.rs:136:28
    |
136 |           std::thread::spawn(move || {
    |  _________------------------_^
    | |         |
    | |         required by a bound introduced by this call
137 | |             file.group("/")
138 | |                 .unwrap()
139 | |                 .new_dataset_builder()
...   |
142 | |                 .unwrap();
143 | |         });
    | |_________^ `*mut u8` cannot be sent between threads safely
    |
    = help: within `CustomType`, the trait `std::marker::Send` is not implemented for `*mut u8`
note: required because it appears within the type `VarLenUnicode`
   --> C:\Users\watsaig\.cargo\git\checkouts\hdf5-rust-c1072d7a2617f7d8\2e44e90\hdf5-types\src\string.rs:288:12
    |
288 | pub struct VarLenUnicode {
    |            ^^^^^^^^^^^^^
note: required because it appears within the type `CustomType`
   --> src\io\src\writer.rs:149:8
    |
149 | struct CustomType {
    |        ^^^^^^^^^^
    = note: required for `Unique<CustomType>` to implement `std::marker::Send`
note: required because it appears within the type `RawVec<CustomType>`

```

Is it possible to make these types thread-safe? I'm happy to help with a PR but not sure where to start.

--- a/hdf5-types/src/array.rs
+++ b/hdf5-types/src/array.rs
@@ -142,6 +142,13 @@ impl<T: Copy + fmt::Debug> fmt::Debug for VarLenArray<T> {
     }
 }
 
+// Safety: `VarLenArray` allocates and frees memory for its data, which is copied
+// from its original location; it never holds a reference to the data used to create it.
+// Therefore if `T` is Send, so is `VarLenArray<T>`
+unsafe impl<T: Copy + Send> Send for VarLenArray<T> {}
+// Safety: No interior mutability or potential for data races if `T` is `Sync`.
+unsafe impl<T: Copy + Sync> Sync for VarLenArray<T> {}
+
 #[cfg(test)]
 pub mod tests {
     use super::VarLenArray;

--- a/hdf5-types/src/array.rs
+++ b/hdf5-types/src/array.rs
@@ -142,11 +142,9 @@ impl<T: Copy + fmt::Debug> fmt::Debug for VarLenArray<T> {
     }
 }
 
-// Safety: `VarLenArray` allocates and frees memory for its data, which is copied
-// from its original location; it never holds a reference to the data used to create it.
-// Therefore if `T` is Send, so is `VarLenArray<T>`
+// Safety: Memory backed by `VarLenArray` can be accessed and freed from any thread
 unsafe impl<T: Copy + Send> Send for VarLenArray<T> {}
-// Safety: No interior mutability or potential for data races if `T` is `Sync`.
+// Safety: `VarLenArray` has no interior mutability
 unsafe impl<T: Copy + Sync> Sync for VarLenArray<T> {}
 
 #[cfg(test)]

--- a/hdf5-types/src/string.rs
+++ b/hdf5-types/src/string.rs
@@ -282,10 +282,9 @@ impl AsAsciiStr for VarLenAscii {
     }
 }
 
-// Safety: `VarLenAscii` allocates and frees memory for its data, which is copied
-// from its original location; it never holds a reference to the data used to create it.
+// Safety: Memory backed by `VarLenAscii` can be accessed and freed from any thread
 unsafe impl Send for VarLenAscii {}
-// Safety: No interior mutability or potential for data races.
+// Safety: `VarLenAscii` has no interior mutability
 unsafe impl Sync for VarLenAscii {}
 
 // ================================================================================
@@ -377,10 +376,9 @@ impl FromStr for VarLenUnicode {
     }
 }
 
-// Safety: `VarLenUnicode` allocates and frees memory for its data, which is copied
-// from its original location; it never holds a reference to the data used to create it.
+// Safety: Memory backed by `VarLenUnicode` can be accessed and freed from any thread
 unsafe impl Send for VarLenUnicode {}
-// Safety: No interior mutability or potential for data races.
+// Safety: `VarLenUnicode` has no interior mutability
 unsafe impl Sync for VarLenUnicode {}
 
 // ================================================================================

--- a/hdf5-types/src/string.rs
+++ b/hdf5-types/src/string.rs
@@ -282,6 +282,12 @@ impl AsAsciiStr for VarLenAscii {
     }
 }
 
+// Safety: `VarLenAscii` allocates and frees memory for its data, which is copied
+// from its original location; it never holds a reference to the data used to create it.
+unsafe impl Send for VarLenAscii {}
+// Safety: No interior mutability or potential for data races.
+unsafe impl Sync for VarLenAscii {}
+
 // ================================================================================
 
 #[repr(C)]
@@ -370,6 +376,12 @@ impl FromStr for VarLenUnicode {
         }
     }
 }
+
+// Safety: `VarLenUnicode` allocates and frees memory for its data, which is copied
+// from its original location; it never holds a reference to the data used to create it.
+unsafe impl Send for VarLenUnicode {}
+// Safety: No interior mutability or potential for data races.
+unsafe impl Sync for VarLenUnicode {}
 
 // ================================================================================
 

